### PR TITLE
[TF-76] file renames + split PythonConvertible protocol

### PR DESCRIPTION
**Pull request by realdoug**
naming & code organization changes as discussed [here](https://github.com/apple/swift/pull/23762#event-2263181050)

### CODE REVIEW

File: test/TensorFlowRuntime/python_conversion.swift [hunk 1]
```diff
@@ -4,12 +4,12 @@
...
 // REQUIRES: tensorflow
 //
-// `numpy.ndarray` conversion tests.
+// `numpy.ndarray` and Python conversion tests.
 
 import TensorFlow
 import StdlibUnittest
 
-var NumpyConversionTests = TestSuite("NumpyConversion")
+var PythonConversionTests = TestSuite("PythonConversion")
```
Comment (dan-zheng) [inline test/TensorFlowRuntime/python_conversion.swift]: I believe you'll need to rename usages of `NumpyConversionTests` to `PythonConversionTests` as well.

File: test/TensorFlowRuntime/python_conversion.swift [hunk 2]
```diff
@@ -4,12 +4,12 @@
 // REQUIRES: swift_test_mode_optimize
 // REQUIRES: tensorflow
 //
-// `numpy.ndarray` conversion tests.
+// `numpy.ndarray` and Python conversion tests.
```
Comment (dan-zheng) [inline test/TensorFlowRuntime/python_conversion.swift]: minor:
```suggestion
// Python conversion and `numpy.ndarray` conversion tests.
```

File: stdlib/public/TensorFlow/PythonConversion.swift [hunk 3]
```diff
@@ -10,8 +10,8 @@
 //
 //===----------------------------------------------------------------------===//
 //
-// This file defines conversion initializers from `numpy.ndarray` to
-// `ShapedArray` and `Tensor`.
+// This file defines conversion logic between Python types &
```
Comment (dan-zheng) [inline stdlib/public/TensorFlow/PythonConversion.swift]:
```suggestion
// This file defines conversions between Python types and TensorFlow library
// types.
```

**Review by dan-zheng — APPROVED** [review]
Thank you! 🙏

**Comment by realdoug** [comment]
@dan-zheng My bad, noticed the Python/Numpy test naming right after hitting submit.  Should be good now.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Review by dan-zheng — APPROVED** [review]

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

File: stdlib/public/TensorFlow/PythonConversion.swift [hunk 4]
```diff
@@ -164,4 +164,21 @@ extension Tensor where Scalar : NumpyScalarCompatible {
...
 }
 
+extension TensorShape : PythonConvertible {
+  public var pythonObject: PythonObject {
+    return dimensions.pythonObject
+  }
+
+  public init?(_ pythonObject: PythonObject) {
+    let hasLen = Bool(Python.hasattr(pythonObject, "__len__"))
+    if(hasLen == true) {
```
Comment (rxwei) [inline stdlib/public/TensorFlow/PythonConversion.swift]:
```suggestion
    if Python.hasattr(pythonObject, "__len__" == true {
```

File: stdlib/public/TensorFlow/PythonConversion.swift [hunk 5]
```diff
@@ -10,8 +10,8 @@
 //
 //===----------------------------------------------------------------------===//
 //
-// This file defines conversion initializers from `numpy.ndarray` to
-// `ShapedArray` and `Tensor`.
+// This file defines conversions between Python types & 
+// custom TensorFlow types
```
Comment (rxwei) [inline stdlib/public/TensorFlow/PythonConversion.swift]: End sentences with a period.

File: stdlib/public/TensorFlow/PythonConversion.swift [hunk 6]
```diff
@@ -1,4 +1,4 @@
-//===-- NumpyConversion.swift ---------------------------------*- swift -*-===//
+//===-- PythonConversion.swift ---------------------------------*- swift -*-===//
```
Comment (rxwei) [inline stdlib/public/TensorFlow/PythonConversion.swift]: This should be 80 columns long.

File: stdlib/public/TensorFlow/PythonConversion.swift [hunk 7]
```diff
@@ -164,4 +164,21 @@ extension Tensor where Scalar : NumpyScalarCompatible {
   public func makeNumpyArray() -> PythonObject { return array.makeNumpyArray() }
 }
 
+extension TensorShape : PythonConvertible {
+  public var pythonObject: PythonObject {
+    return dimensions.pythonObject
+  }
+
+  public init?(_ pythonObject: PythonObject) {
```
Comment (rxwei) [inline stdlib/public/TensorFlow/PythonConversion.swift]: I didn't review the last change closely. Maybe tests broke because `TensorShape(0)`'s initialization defaulted to `TensorShape(0 as PythonObject)` somehow?

File: stdlib/public/TensorFlow/PythonConversion.swift [hunk 8]
```diff
@@ -164,4 +164,21 @@ extension Tensor where Scalar : NumpyScalarCompatible {
...
+    let hasLen = Bool(Python.hasattr(pythonObject, "__len__"))
+    if(hasLen == true) {
+      guard let array = [Int32](pythonObject) else { return nil }
+      self.init(array)
+    } else {
+      guard let num = Int32(pythonObject) else { return nil }
```
Comment (rxwei) [inline stdlib/public/TensorFlow/PythonConversion.swift]: Matching scalar does seem right to me. Why is this necessary?

**Review by rxwei — CHANGES_REQUESTED** [review]
There's some important issues I missed in the last merged PR.

File: stdlib/public/TensorFlow/PythonConversion.swift (hunk 6 above)
Comment (realdoug) [reply stdlib/public/TensorFlow/PythonConversion.swift]: @rxwei is there a linter I can install to catch this kind of stuff?

File: stdlib/public/TensorFlow/PythonConversion.swift (hunk 7 above)
Comment (realdoug) [reply stdlib/public/TensorFlow/PythonConversion.swift]: @rxwei Yep, that is exactly the case.  I spent some time looking into how to manipulate the  prority order of the initializers, but if you have a suggested approach i can definitely try it out.  I settled for just making sure that this initializer successfully passes through to the correct initializer, since ```TensorShape(0 as PythonObject)``` should be valid anyway.

It is also possible that the failure was due to this initializer being failable, but the only reliable repro steps are what I outlined [here](https://bugs.swift.org/browse/TF-426).

I guess the question is what would you want/expect the output of the below program to be?
```
func doThing(_ arg: PythonObject) {
  print(type(of: arg))
}

func doThing(_ arg: Int32) {
  print(type(of: arg))
}

doThing(1)
```

Output currently is PythonObject.  Perhaps that is the real underlying bug.

File: stdlib/public/TensorFlow/PythonConversion.swift (hunk 8 above)
Comment (realdoug) [reply stdlib/public/TensorFlow/PythonConversion.swift]: @rxwei the reason is that ```[Int32](1)``` (and thus ```[Int32](1 as PythonObject)```) is not valid in Swift, but ```TensorShape(1)``` is valid.  Or do you mean that you would prefer ```[Int32]([pythonObject])``` here?

File: stdlib/public/TensorFlow/PythonConversion.swift (hunk 7 above)
Comment (rxwei) [reply stdlib/public/TensorFlow/PythonConversion.swift]: I see. There are two bigger problems. First, the TensorFlow module shouldn't logically depend on Python, i.e. there can be targets that require all APIs that use Python be `#if`'d out. Second, we ultimately want to optimize tensor operations through techniques like graph program extraction, but having this Python layer for the initialization of a simple tensor shape makes it extremely inefficient when we are targeting hardware like TPUs.

I think a better solution would be to split this protocol into two. This has actually been on my mind for over a year now:

`PythonConvertible` defines conversion **to** Python; `ConvertibleFromPython` defines conversion **from** Python.

`TensorShape` will then only conform to `PythonConvertible`, but not the other one, making this integer literal issue go away.

Comment (realdoug) [reply stdlib/public/TensorFlow/PythonConversion.swift]: Ah, makes sense.  I could definitely work on a PR for the split up of the protocol.

Comment (rxwei) [reply stdlib/public/TensorFlow/PythonConversion.swift]: Great. Let's do that!

**Comment by realdoug** [comment]
@dan-zheng ok, made the syntax updates here in case you want to merge those.  I'll follow up with the protocol split as discussed later this weekend.

**Comment by realdoug** [comment]
@dan-zheng @rxwei this branch has the protocol split on it now.

**Comment by rxwei** [comment]
Oh excellent. We can skip [the revert](https://github.com/apple/swift/pull/24044) then.

File: stdlib/public/Python/Python.swift [hunk 9]
```diff
@@ -1071,14 +1097,16 @@ extension Range : PythonConvertible where Bound : PythonConvertible {
...
+extension PartialRangeFrom : PythonConvertible where Bound : PythonConvertible {
   public var pythonObject: PythonObject {
     _ = Python // Ensure Python is initialized.
-    return Python.slice(lowerBound, upperBound, Python.None)
+    return Python.slice(lowerBound, Python.None, Python.None)
   }
 }
 
-extension PartialRangeFrom : PythonConvertible where Bound : PythonConvertible {
+extension PartialRangeFrom : ConvertibleFromPython where Bound : ConvertibleFromPython {
```
Comment (rxwei) [inline stdlib/public/Python/Python.swift]: Fit within 80 columns.
```suggestion
extension PartialRangeFrom : ConvertibleFromPython
  where Bound : ConvertibleFromPython {
```

File: stdlib/public/Python/Python.swift [hunk 10]
```diff
@@ -1088,14 +1116,16 @@ extension PartialRangeFrom : PythonConvertible where Bound : PythonConvertible {
...
+extension PartialRangeUpTo : PythonConvertible where Bound : PythonConvertible {
   public var pythonObject: PythonObject {
     _ = Python // Ensure Python is initialized.
-    return Python.slice(lowerBound, Python.None, Python.None)
+    return Python.slice(Python.None, upperBound, Python.None)
   }
 }
 
-extension PartialRangeUpTo : PythonConvertible where Bound : PythonConvertible {
+extension PartialRangeUpTo : ConvertibleFromPython where Bound : ConvertibleFromPython {
```
Comment (rxwei) [inline stdlib/public/Python/Python.swift]: Fit within 80 columns.
```suggestion
extension PartialRangeUpTo : ConvertibleFromPython
  where Bound : ConvertibleFromPython {
```

File: stdlib/public/Python/Python.swift [hunk 11]
```diff
@@ -1042,26 +1073,21 @@ extension Dictionary : PythonConvertible
...
-      Py_DecRef(k)
-      Py_DecRef(v)
-    }
-    return PythonObject(consuming: dict)
-  }
 }
 
 //===----------------------------------------------------------------------===//
-// `PythonConvertible` conformance for `Range` types
+// `PythonConvertible` and `ConvertibleFromPython` conformance
```
Comment (rxwei) [inline stdlib/public/Python/Python.swift]:
```suggestion
// `PythonConvertible` and `ConvertibleFromPython` conformances
```

**Review by rxwei — APPROVED** [review]
This is awesome. Thanks Doug!

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
CI seems dead. This PR needs to go in before https://github.com/apple/swift/pull/24012. @dan-zheng could you help make sure this gets tested?

File: stdlib/public/Python/Python.swift [hunk 12]
```diff
@@ -973,6 +979,16 @@ extension Float : PythonConvertible {
...
+  public var pythonObject: PythonObject {
+    return self?.pythonObject ?? Python.None
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// `ConvertibleFromPython` conformance for `Optional`
+//===----------------------------------------------------------------------===//
+
+extension Optional : ConvertibleFromPython where Wrapped : ConvertibleFromPython {
```
Comment (dan-zheng) [inline stdlib/public/Python/Python.swift]:
```suggestion
extension Optional : ConvertibleFromPython
  where Wrapped : ConvertibleFromPython {
```

**Review by dan-zheng — APPROVED** [review]

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow, seriously

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow linux

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by dan-zheng** [comment]
There is one test failure:
```
[ RUN      ] PythonConversion.shaped-array-conversion
stdout>>> check failed at /Volumes/BuildData/Users/swiftci/jenkins/workspace/swift-PR-TensorFlow-macOS/swift/test/TensorFlowRuntime/python_conversion.swift, line 62
stdout>>> expected: (2, 7, 2) (of type Python.PythonObject)
stdout>>> actual: [2, 7, 2] (of type Python.PythonObject)
stdout>>> check failed at /Volumes/BuildData/Users/swiftci/jenkins/workspace/swift-PR-TensorFlow-macOS/swift/test/TensorFlowRuntime/python_conversion.swift, line 64
stdout>>> expected: (14, 2) (of type Python.PythonObject)
stdout>>> actual: [14, 2] (of type Python.PythonObject)
[     FAIL ] PythonConversion.shaped-array-conversion
[ RUN      ] PythonConversion.tensor-conversion
[       OK ] PythonConversion.tensor-conversion
[ RUN      ] PythonConversion.tensor-round-trip
[       OK ] PythonConversion.tensor-round-trip
[ RUN      ] PythonConversion.tensor-shape
[       OK ] PythonConversion.tensor-shape
PythonConversion: Some tests failed, aborting
UXPASS: []
FAIL: ["shaped-array-conversion"]
SKIP: []
```

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by realdoug** [comment]
ugh sorry bout that!  Could have sworn that passed locally for me.
